### DISCUSSION TITLE
WebRTC-HTTP ingestion protocol (WHIP)

### DISCUSSION THREAD

**Discussion by JonasPrimbs** [OvenMediaLabs/OvenMediaEngine#1096] [Ideas]
Hey all,

I was just taking a look at your implementation of the WebRTC upstream endpoint and in my opinion the "protocol" is a bit to overengineered.

Most of it (except the ICE trickling) can be done unsing one simple POST request for startup:
The request contains the Session Description and ICE Candidates of the client and the response contains the ones of the server. (This could be divided into two separate requests.)
The connection can then be closed using a simple HTTP DELETE request.
And that's it! No WebSocket connection, no state management, no issues when the WebSocket connection eventually breaks! Just a simple, minimalistic REST API.

And, here comes the best, this corresponds to a current (and very early) IETF draft, called "WebRTC-HTTP ingestion protocol (WHIP)": https://datatracker.ietf.org/doc/draft-murillo-whip/
So, why not trying to take part at the standardization process and make the WebRTC implementation more future-proof and interoperable to other tools?

By the way: I'm not part of this IETF working group. I'm just interested in the WebRTC technology and watching the draft for some months and it will make me happy to see it in action in my favorite streaming backend.

**Comment by dimiden** [comment]
@JonasPrimbs 
Thank you for your opinion!

As you mentioned, WebRTC signalling can be implemented in many other ways besides WebSocket, and the reason why we used WebSocket as a signalling method is to easily expand the functionality.

A typical example is the [P2P Delivery](https://app.gitbook.com/@airensoft/s/ovenmediaengine/p2p-delivery) feature.
Regular WebRTC playback is sufficient for REST APIs, but connection-oriented protocols such as WebSocket are beneficial for SDP negation between host/client peers.

Lastly, since I don't know much about WHIP, I think I need to review it with the OME developers.

**Comment by getroot** [comment]
WHIP has been mentioned several times when developing the WebRTC Provider. I'm reviewing this very positively, and will introduce it to OME someday. However, it is not an easy task as the structure of the current OvenMediaEngine needs to be changed as mentioned below. When I finish my high-priority task, I will start designing for WHIP.

https://github.com/AirenSoft/OvenMediaEngine/issues/29#issuecomment-826784958

**Comment by alexthedamager** [comment]
This looks interesting - WebRTC coming to vanilla OBS, but possibly not using WHIP? https://github.com/obsproject/rfcs/pull/43

**Comment by alexthedamager** [comment]
OBS recently stated that native WebRTC support in OBS is due Q3 of 2022:
https://github.com/obsproject/obs-studio/discussions/4021#discussioncomment-2127284 

😳

[a comment by an automated account is not shown]

**Comment by alexthedamager** [comment]
OBS has just released a prototype / playground for WebRTC / WHIP support within OBS: https://github.com/obsproject/obs-studio/pull/7192

**Comment by alexthedamager** [comment]
Update: OBS now has an [open PR](https://github.com/obsproject/obs-studio/pull/7926) for WebRTC / WHIP output. 

An easy way to test this for those curious is by using broadcast-box: https://b.siobud.com/
https://github.com/Glimesh/broadcast-box

Exciting times!

**Comment by getroot** [comment]
@llspalex 
I will start developing WHIP soon. So I am collecting material for this. I checked the WHIP RFC and the PR for OBS you posted. But I need a binary or installer of OBS for Windows with WHIP support to make things go smoothly, but I haven't been able to find one. do you know where I can get this? Or should I compile and use OBS myself?

**Comment by nums** [comment]
@getroot you can also use this project : https://github.com/meetecho/simple-whip-client

**Comment by getroot** [comment]
@nums Thank you. I've never used Gstreamer yet, so I'm a bit confused. But I think it will be very useful for WHIP development.

**Comment by alexthedamager** [comment]
@getroot 
The x64 Windows build in the "artifacts" section on the bottom of this page should have WHIP support: https://github.com/obsproject/obs-studio/actions/runs/3969201209 

Also, here is a link to the latest WebRTC PR for OBS with recent discussion (it is not the same PR as the original "playground" PR I posted): https://github.com/obsproject/obs-studio/pull/7926

**Comment by getroot** [comment]
@llspalex 
Thank you. I downloaded OBS from the link you provided and confirmed that WHIP exists. 

I was already checking the [obsproject/obs-studio#7926](https://github.com/obsproject/obs-studio/pull/7926) PR. I felt that OBS will officially support WHIP soon, so I will start working on it soon.

**Comment by getroot** [comment]
I've committed WHIP support to the master branch.
Please see #1043 for further discussion of WHIP.
